### ISSUE TITLE
Inaccurate predictions shear 2 point correlation functions at small scales

### ISSUE THREAD

**Issue by JegerBroxterman** [cloe-org/cloelib#266]
## 📝 Issue Summary
I am using` cloelib 0.8.2` with the playground notebook at `/playground/tutorials/observables/photo.ipynb`. The notebook runs fine, but if I extend the prediction of the shear 2PCFs to smaller scales than shown in the notebook, the predictions seem wrong (see plots attached). Maybe some integration limits are too strict, so the smallest scales are not predicted correctly. Changing `ells_integration = np.arange(2, 30000)` in the notebook does not solve the issue. I also attached a prediction from my own code that has a smoother prediction for much smaller angles to show what I would expect.

### 🎯 Goals / Acceptance Criteria
- [ ] Identify origin of unexpected behaviour.
- [ ] Fix issue to allow for accurate predicition at small scales

### 🔄 Steps to Reproduce (if applicable)
1. Git clone cloelib and cloe playground
2. In /playground/tutorials/observables/photo.ipynb change angular range to predicht xipm to 
```
# Define at which theta to evaluate the correlation function
theta_deg = np.geomspace(0.1, 100, 20)/60 # in degrees
theta_rad = np.deg2rad(theta_deg)  # in radians
```
3. Run notebook 

### ✅ Expected Outcome
Accurate predicition of shear two-point correlation function at small scales. 

### 🐞 Current Behavior (if applicable)
Cloe outcome:
<img width="960" height="540" alt="Image" src="https://github.com/user-attachments/assets/2de553b7-f989-4a90-aa8e-a40754cff929" />
Note that there are multiple data points at the flat parts at theta < 10e-2 deg.

Expected:
<img width="960" height="540" alt="Image" src="https://github.com/user-attachments/assets/dc5ec9a9-7e12-4c25-92c2-568b8a9c5bb9" />

### 📌 Additional Notes
The angular power spectra / C_ells visually seem ok to me.

### 👥 Assignees
- @TBA

**Comment by pburger112** [comment]
Hi, I see the same issue. But in my case, it does change with ell_max. But it never gets pretty accurate. And increasing ell_max>40k takes for ever

**Comment by JegerBroxterman** [comment]
@pburger112 is right that the smallest scales indeed change depending on what ell range is set, but for me, changing it also still doesn't solve the issue.

**Comment by gcanasherrera** [comment]
This is clearly a bug. Being a bit cheeky (but kind, nice and _grateful_), @llinke1 would you have maybe 20 min to have a look at this?

**Comment by gcanasherrera** [comment]
After investigations with @marcobonici , we have concluded that this bug requires further attention as some follow up problems have been observed. The feature that @JegerBroxterman and @pburger112 are seeing is mostly coming from the poorly sampled theta array, but further investigations are necessary.

**Comment by gcanasherrera** [comment]
Milestone moved to v0.10

**Comment by gcanasherrera** [comment]
There are also further issues with the correlation function calculation:

- Low speed despite the caching mechanism

I know @marcobonici was doing tests

**Comment by gcanasherrera** [comment]
@ippoppi, do you mind reporting here the accuracy issue? We can get this fixed and then open up a follow up for speed

**Comment by ippoppi** [comment]
> [@ippoppi](https://github.com/ippoppi), do you mind reporting here the accuracy issue? We can get this fixed and then open up a follow up for speed

Hi @gcanasherrera, the issue is caused by insufficient numerical precision in the computation of the Wigner d-matrices. Increasing the precision by using `float64` instead of `float32` solves the problem (see plot). Alternatively, implementing a different formula for the Wigner d-matrix computation could also solve the issue.

Best,

Filippo

<img width="761" height="553" alt="Image" src="https://github.com/user-attachments/assets/80dcdfe7-92a9-4f2c-846b-440b1d5c8232" />

**Comment by marcobonici** [comment]
Hi @ippoppi , this is great! Regarding the new formula: do you already have something in mind? Using a different stable-recurrence?

**Comment by gcanasherrera** [comment]
If not, we could potentially just fix the precision (amazing job @ippoppi) and work out speed/improvements in a follow up test?

In this way, users can still use correlation functions without the bug to make theoretical predictions (although slower).

**Comment by ahallcosmo** [comment]
I see (thanks to @pburger112) that the wigner_d computation switches to using an approximation (`approximation_fn`) for `ell>30000`, which doesn't look like it will be accurate at small theta. An alternative would be to switch to the asymptotic 'flat sky' formula above some threshold, i.e., J0(ell \* beta) for d22 and d00, J4(ell \* beta) for d2-2, and J2(ell \* beta) for d20. This is what, e.g., GSL does for the Legendre polynomials (where a threshold of L=100000 is used).

**Comment by ippoppi** [comment]
> Hi @ippoppi , this is great! Regarding the new formula: do you already have something in mind? Using a different stable-recurrence?

Hi @marcobonici , it should be possible to approximate the Wigner d-matrices with Bessel functions, as @ahallcosmo already noted. However, I don't know how this would perform in terms of precision and computational time, or if it is immediately jit compatible without further approximation. Unless we're using incredibly small angles, I think increasing the precision should suffice.

Best,

Filippo

**Comment by gcanasherrera** [comment]
Hi all,

Thanks for all the feedback everyone! A few thoughts on my side on how I see this:

**Urgency**: Right now, the priority is to have a correct computation of the correlation functions so we can continue building on top (blinding shifts, likelihood). If increasing the accuracy works, let’s go with that.
**Next**: Most likely, the next focus will be speed. We have two routes to explore:

- Further optimize @marcobonici’s magic with the cache mechanism.
- Continue brainstorming in the direction of @ahallcosmo switch / @ippoppi Bessel functions.

Shall we close this issue by increasing the accuracy and move the discussion about speed to another follow-up, so we can start closing concerns? :)

Good job, everyone!

**Comment by gcanasherrera** [comment]
@all-contributors please add @pburger112 for bug
@all-contributors please add @JegerBroxterman for bug
@all-contributors please add @ahallcosmo for ideas
@all-contributors please add @ippoppi for ideas

**Comment by gcanasherrera** [comment]
@all-contributors please add @JegerBroxterman for bug

**Comment by gcanasherrera** [comment]
@all-contributors please add @ahallcosmo for ideas

**Comment by ippoppi** [comment]
Hi,

Sorry for the lengthy message. I have studied how to increase the precision and also implemented the Bessel approximation. Here's a summary of what I've come up with.

The 64-bit precision is enabled by a global jax configuration that also changes the default precision, triggered by 
`jax.config.update("jax_enable_x64", True)`.
This makes **all** the jax variables defined after this call 64-bits, unless explicitly defined as `float32` or `int32`(and not simply as `float` or `int`).
For example, if we add this option in _angular_correlation_function_wigner.py_, all the jax variables defined after its import will become 64-bits
So that, we could have a situation like the one in this example: 

```
import jax.numpy as jnp
t=jnp.array([0.]) #this is a float32
from cloelib.summary_statistics.angular_correlation_function_wigner import  AngularCorrelationFunctionWigner # this contains jax.config.update("jax_enable_x64", True)
r=jnp.array([0.]) #this is a float64
```
This is clearly an unwanted behavior since we have type mixing, and the user does not have clear control of the types.
Note that jax is coded like that on purpose, to enforce the use of a single dtype in all the code. This option is expected to be called at the top of the main file (see the [documentation](https://docs.jax.dev/en/latest/default_dtypes.html) ) 

If we choose this road, we must be careful in how to implement it. 
This is not really a coding problem, but is more related to the philosophy behind the library (i.e., it should be faster or more precise, ready to use or highly personalizable, etc.)

I have thought about some options:

1. Turn the whole code into 64-bits. This is the safer option, but it will slow down all the computation (not dramatically) and will increase the memory usage.  Personally, I always prefer to work in double precision (better safe than sorry).
2.  Letting the user set the precision with a dedicated function. Maybe a warning could appear when the user tries to use small angles, recommending increasing the precision. 
3. As in the previous one, leaving the choice to the user but setting the default to 64 and enabling an optional "fast mode" that decreases the precision to 32.
4. Explicitly specify the precision in all the jax functions (so changing all the `float` to `float32` or `float64`, and the same with the `int`). In this case, we could also create multiple versions of the same function called conditionally. Quite complex to implement, the entire library would need to be updated, making the process very prone to error.

Among these, my personal preference would probably be the option 3.

Alternatively, we can compute the d matrix with the Bessel function approximation under the given condition.
I could not find any pure jax implementation of the J, so I did it with a callback of the scipy `jv` function. 
It is way slower, but it seems to work and can be implemented, leaving the rest of the library untouched.

Let me know what you think. We could also have a call to discuss these points in more detail.

Best, Filippo

**Comment by ahallcosmo** [comment]
@ippoppi I'm also in favour of having 64-bit as the default precision.

Regarding the switch over to Bessel functions, another option is to note that we only ever need Wigner d to compute correlation functions from Cls (we don't compute angular correlation function covariances in CLOE, I think?), and in the flat sky approximation the Hankel transform can be performed with FFTLog. So we could just avoid explicitly computing the Bessel function and build the flat-sky approximation into the correlation function calculation at small theta.

**Comment by gcanasherrera** [comment]
Hi @ippoppi many thanks for this! All of this is absolutely great.

Few thoughts on my side:
- We have your amazing profiling functions. Can we quickly check how the speed changes, in particular for the angular power spectra, switching from float32 to float64? I would like to grasp how slower the likelihood would become.
- In any case, I agree, I think option 3 is the way to go (@marcobonici, confirm?) 

I know @davidesciotti was thinking about calculation covariances for real space using cloelib. How are we doing with that? Potentially, switching to flat-sky approximation at small theta could be the way to go, although I would like to think a bit more about this, as I can't believe (I know you're probably right @ippoppi ) there is any Jax implementation of J just yet.

**Comment by ippoppi** [comment]
Hi,

@ahallcosmo thanks for the suggestion!

@gcanasherrera I just profiled the `get_Cl` function, and it actually runs slightly faster with double precision. I suspect this anomalous behavior is caused by mixing jax 32 bit outputs with numpy 64 bit variables. The overhead from these implicit type conversions takes more time than the jax computation. I don't think this will be the case for other functions; but we should check them one by one to verify. However, unless the choice of single precision was made with some specific application in mind (like GPUs), I think that upgrading all the jax functions to 64-bit is the best choice.

Regarding the Js implementation in jax, I could not find it and I tried the faster way to implement it to verify if it was feasible. However I'm no jax expert, and I didn't have much time to search for it either. So, my conclusion is by no means definitive. If implementing double precision causes any issues, I'll investigate further.

best,

Filippo

**Comment by gcanasherrera** [comment]
Many thanks @ippoppi for all your work! Let's change the precision to 64-bits, this is the most sensible decision!

**Comment by davidesciotti** [comment]
@gcanasherrera and @ahallcosmo just to clarify, no, we never compute any covariances in CLOE, and the current real-space covariance projection (in Spaceborne) is built independently from cloelib...
